### PR TITLE
created websockets and toast popup for the recently finished meetings

### LINKED ISSUE #167: Use WebSockets and toast popups to inform users of recently finished meetings
Use WebSockets and toast popups to inform users of recently finished meetings

### PR DESCRIPTION
**Pull request by subhangi2731**
Using socket.io to send and receive data across browser and backend server. The function pings the endpoint every 5 seconds to fetch meetings that have been completed in the last 60 seconds and thus informing users for the recently finished meetings

Closes  #167

### CODE REVIEW

File: src/mote/templates/mainpage.html
Comment (darknao) [inline src/mote/templates/mainpage.html:416]: If you listen to the 'connect' event, you don't need to send any additional messages, and can remove this block.

File: src/mote/main.py
Comment (darknao) [inline src/mote/main.py:147]: Try to use a meaningful name, like on_connect or something like that.

File: src/mote/templates/mainpage.html
Comment (darknao) [inline src/mote/templates/mainpage.html:402]: I think we need a little more information in this notification. The meeting name at least.

File: src/mote/templates/mainpage.html
Comment (darknao) [inline src/mote/templates/mainpage.html:413]: I don't think you need to pass any arguments if you're using the same server
```suggestion
      var socket = io.connect()
```

**Review by darknao — CHANGES_REQUESTED** [review]
Ok, so this is a little bit suboptimal:
You are starting a new thread every time a client connect.
Each threads will poll datagrepper, and broadcast the result to all connected clients.

Let say we have 5 clients connected, you end up with 5 separates threads that does the exact same thing.
When a meeting end, each thread will send a message to **all** clients (in this case, each client will receive 5 times the same message).
And even worst: those threads are never terminated, even when the client disconnect.
This can cause some serious amplification attack toward the datagrepper service, and we don't want that :)

A better approach would be to only start one thread when at least one client is connected, and stop it when the last client disconnect.

**Comment by subhangi2731** [comment]
@darknao I've updated the PR with all the changes mentioned above. 
The updated PR has the following changes:
* Each time a client connects, the program logic checks if the function for pulling data from Datagrepper endpoint is already running. If it does, it does not run the function again.
* The single running function now broadcasts the toast message to all of the running clients
* When the clients are connected, a counter is kept to maintain a record of the number of clients connected at once. This helps close the running function once all the clients are disconnected. In short, no more API calls once the clients are all disconnected, hence saving resources.
* The function is called when atleast one client connects, and is stopped when the last one disconnects.
* Redundant messages used to handle socket.io events are now removed.
* Function names are made more meaningful
* Toast notification now even has the name of the finished meeting, along with a `Open meeting`  hyperlink and a `Close` button for the toast to close.
* When there are multiple meetings finished in the last minute, the code now shows up a toast for each of them, in a span of `5 seconds` so that the user gets time to react to each toast.

new

**Comment by subhangi2731** [comment]
@darknao @t0xic0der I did all the changes whatever told please check and let me know

File: src/mote/main.py
Comment (darknao) [inline src/mote/main.py:157]: Few recommendations:
Use thread lock mechanism to avoid race condition.
For the first client connected, this function is blocked until the `fetch_recently_completed_meetings` exit.
Start a background task instead. You can use the socketIO utility function `socketio.start_background_task` for that purpose.

File: src/mote/main.py
Comment (darknao) [inline src/mote/main.py:166]: When the last client disconnects, there is a 60 seconds period when `is_running` is False, but the while loop is still running the sleep call.
If another client connects during this time, it starts a new loop, and the previous one keeps running.
You need to make sure the loop is stopped before running a new one.

File: src/mote/templates/mainpage.html
Comment (darknao) [inline src/mote/templates/mainpage.html:428]: Instead of doing this one by one, maybe we can stack notifications ?

**Review by darknao — CHANGES_REQUESTED** [review]
You can take inspiration from this example code from socketIO to see how to handle background task in a cleaner way : https://github.com/miguelgrinberg/Flask-SocketIO/blob/main/example/app.py

**Comment by subhangi2731** [comment]
I have implemented stacking of toasts currently, working on implementing background task now.

File: src/mote/main.py
Comment (subhangi2731) [reply src/mote/main.py:154]: Used socketio's background task now, instead of manual triggering through the use of global variables.

**Comment by subhangi2731** [comment]
@darknao I have made all the changes. 
* Socketio's background task feature is now used to start the function
* A threading Event now terminates the function when set to `True` despite any sleep/wait cycle.

File: src/mote/main.py
Comment (darknao) [inline src/mote/main.py:149]:
```suggestion
        exit.wait(60)
```

File: src/mote/main.py
Comment (darknao) [inline src/mote/main.py:50]: `exit` is already used internally by Python, choose another name, like `exit_thread`.

**Review by darknao — CHANGES_REQUESTED** [review]
Ok, that's nice.
I still have only one notification showing after another, instead of all of them at once.
And maybe we should leave them a bit longer ? 5 seconds seems short to me.

**Comment by subhangi2731** [comment]
@darknao 
* I have removed wait time when showing toasts, since that is the reason for toasts being displayed individually and not in a stack.
* I have updating with better naming conventions.

**Comment by subhangi2731** [comment]
@t0xic0der @darknao please review my updated PR thanks!

**Review by darknao — APPROVED** [review]

**Comment by darknao** [comment]
Seems good to me :+1: 
I still think the 5 seconds display time is a bit too fast though...

**Comment by subhangi2731** [comment]
@darknao Should I decrease the display time?

**Comment by darknao** [comment]
> @darknao Should I decrease the display time?
I mean increase it a little. Like 8 or 10 seconds maybe?

**Comment by subhangi2731** [comment]
@darknao sure

**Comment by subhangi2731** [comment]
@darknao can you please point me out where do I need to increase the display time as the 5-second time has already been removed, because it led to single toasts being shown instead of in stack

File: src/mote/templates/mainpage.html
Comment (darknao) [inline src/mote/templates/mainpage.html:427]: One way of increasing the delay (in ms):
```suggestion
                var toast = new bootstrap.Toast(div, {delay: 10000})
```

**Comment by t0xic0der** [comment]
@subhangi2731, please provide us with updates on this if there are any.

**Comment by subhangi2731** [comment]
@t0xic0der @darknao sure please check and let me know

**Comment by darknao** [comment]
Looks good.
Please rebase your branch and solve conflicts, then it should be good to merge.

**Comment by t0xic0der** [comment]
@darknao, could you please provide me with the steps you performed to review the PR - function-wise?

**Comment by subhangi2731** [comment]
@darknao okay I will do it

**Comment by darknao** [comment]
> @darknao, could you please provide me with the steps you performed to review the PR - function-wise?

Old fashioned manual testing using logging and playing with the call_period of datagrepper for python side, and dev console for the js side.

**Comment by subhangi2731** [comment]
@darknao This has no conflicts, so what else am I expected to do here as I resolved everything and all the checks passed

**Comment by t0xic0der** [comment]
@darknao, aren't this and https://github.com/fedora-infra/mote/pull/218 ending up doing the same thing? I'm curious.

**Comment by darknao** [comment]
#218 is based on this PR (they share the same commit history)
This one is adding the notification part using websocket and datagrepper for fetching recent meetings.
The other one replace the datagrepper part with fedora-messaging.

**Comment by t0xic0der** [comment]
Makes sense. I'll give this PR another look once I have an merge approval from your end.

File: src/mote/main.py
Comment (t0xic0der) [inline src/mote/main.py:25]: I don't see the `time` library being used here. Please remove its import.

File: src/mote/templates/mainpage.html
Comment (t0xic0der) [inline src/mote/templates/mainpage.html:435]: This is nitpicking but can you please make the indentation uniform for the JavaScript block?

File: src/mote/templates/mainpage.html
Comment (t0xic0der) [inline src/mote/templates/mainpage.html:36]: Please download and store the JavaScript library in `src/mote/static/jscn/` and reference it from there - just like we're doing for other such JavaScript dependencies.

**Review by t0xic0der — CHANGES_REQUESTED** [review]
Some more suggested changes.

**Comment by t0xic0der** [comment]
I added https://github.com/fedora-infra/mote/pull/201/commits/916b69a29ba76fb8951e6fdf88496da6ee49067c. Please be sure to pull before making any changes.

**Comment by subhangi2731** [comment]
@t0xic0der sure I will do this

**Comment by subhangi2731** [comment]
@t0xic0der I resolved the indentation and removed time import but where is the js source file for socket io I can't find it in the website can you please tell

File: src/mote/templates/mainpage.html
Comment (t0xic0der) [inline src/mote/templates/mainpage.html:36]: As a reply to https://github.com/fedora-infra/mote/pull/201#issuecomment-982374808, the source is present in this [line](https://github.com/fedora-infra/mote/pull/201/files#diff-4c11e602a99bd1a8ca037ca1be41a6574c9c32ce26b2a7bee39fdd0c46fec471R36).

Comment (subhangi2731) [reply src/mote/templates/mainpage.html:36]: @t0xic0der yes yes I got it thanks

**Comment by subhangi2731** [comment]
@t0xic0der @darknao I rebased the branch resolved all the conflicts that came in the way (mainpage.html and main.py) and did all the changes as said and added the dependency please check and let me know

**Comment by darknao** [comment]
looks good to me.
I don't see the commit from @t0xic0der anymore, but it seems to be included in your commit now, so I guess it's ok.
There is still a merge commit in the history, which mean a rebase is needed.

**Comment by subhangi2731** [comment]
@darknao yes due to the conflicts it created an issue but I downloaded the dependency and included it. Ok I will just do the rebase thing

**Comment by subhangi2731** [comment]
@darknao @t0xic0der  please have a look now

**Comment by darknao** [comment]
Yep, that's fine for me :+1:

File: src/mote/templates/mainpage.html
Comment (t0xic0der) [inline src/mote/templates/mainpage.html:435]:
```suggestion
        var socket = io.connect();
        function showToast(meeting_json) {
            meeting_count = meeting_json.length;
            const timer = ms => new Promise(res => setTimeout(res, ms));
            async function load() {
		        for (var i = 0; i < meeting_count; i++) {
		            var toastDiv = $(`
		                <div class="toast-body">
		                    <h6 id="meeting-name">`+ meeting_json[i]["msg"]["meeting_topic"] +`</h6>
		                    <div class="mt-2 pt-2 border-top">
		                        <a id="toast-btn" type="button" class="btn btn-primary btn-sm" href="`+ meeting_json[i]["msg"]["url"] +`" target="_blank">Open meeting</a>
		                        <button type="button" class="btn btn-secondary btn-sm" data-bs-dismiss="toast">Close</button>
		                    </div>
		                </div>
		            `);
		            var parentDiv=document.getElementById('parent');
		            var div = document.createElement('div');
		            div.setAttribute('class','toast');
		            div.setAttribute('role','alert');
		            div.setAttribute('aria-live','assertive');
		            div.setAttribute('aria-atomic','true');
		            $(div).append(toastDiv);
		            $(parentDiv).append(div);
		            var toast = new bootstrap.Toast(div, {delay: 10000});
		            toast.show();
		        }
        	}
        	load();
        }
        socket.on("show_toast", function(meeting_json) {
            showToast(meeting_json);
        });
```

**Review by t0xic0der — CHANGES_REQUESTED** [review]
Oh I so passionately hate the ES6 inline functions!

Anyway, here are some suggested changes.

**Comment by subhangi2731** [comment]
@t0xic0der done

**Comment by t0xic0der** [comment]
@subhangi2731, I noticed that you amended the most recent commit https://github.com/fedora-infra/mote/commit/9443aa618495395abcc4680214bc5107ac7adefb to accommodate the changes that I suggested on https://github.com/fedora-infra/mote/pull/201#pullrequestreview-819783983 even though the changes are disjointed from the commit message. Please try to keep like changes together and if needed, you can always create a new commit to address unlike changes rather than pitting together the unlike ones.

**Review by t0xic0der — APPROVED** [review]
Sigh.

LGTM.

**Comment by subhangi2731** [comment]
@t0xic0der sure I will take care of this from the next time. The problem happened as there were many conflicts I had to rebase the branch two times to fix the commits. I thought to include all commit messages in one but I will create a new commit the next time for the particular changes sure

**Pull request merged**
